x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/details/index.tsx

```diff
@@ -241,24 +251,43 @@ const RuleDetailsPageComponent: React.FC<DetectionEngineComponentProps> = ({
           defineRuleData: null,
           scheduleRuleData: null,
         };
+  const [dataViewTitle, setDataViewTitle] = useState<string>();
+  useEffect(() => {
+    const fetchDataViewTitle = async () => {
+      if (defineRuleData?.dataViewId != null && defineRuleData?.dataViewId !== '') {
+        const dataView = await data.dataViews.get(defineRuleData?.dataViewId);
+        setDataViewTitle(dataView.title);
+      }
+    };
+    fetchDataViewTitle();
+  }, [data.dataViews, defineRuleData?.dataViewId]);
   const [showBuildingBlockAlerts, setShowBuildingBlockAlerts] = useState(false);
   const [showOnlyThreatIndicatorAlerts, setShowOnlyThreatIndicatorAlerts] = useState(false);
   const mlCapabilities = useMlCapabilities();
   const { formatUrl } = useFormatUrl(SecurityPageName.rules);
   const { globalFullScreen } = useGlobalFullScreen();
   const [filterGroup, setFilterGroup] = useState<Status>(FILTER_OPEN);
+  const [dataViewOptions, setDataViewOptions] = useState<{ [x: string]: DataViewListItem }>({});
 
+  useEffect(() => {
+    const fetchDataViews = async () => {
+      const dataViewsRefs = await data.dataViews.getIdsWithTitle();
+      if (dataViewsRefs.length > 0) {
+        const dataViewIdIndexPatternMap = dataViewsRefs.reduce(
+          (acc, item) => ({
+            ...acc,
+            [item.id]: item,
+          }),
+          {}
+        );
+        setDataViewOptions(dataViewIdIndexPatternMap);
+      }
+    };
+    fetchDataViews();
+  }, [data.dataViews]);
```

Comment:
Similar to below -- does this need to happen here, or can it be pushed further down into `StepDefineRule`? Maybe even all the way to the read-only viewer component?

Reply:
Yeah this is interesting. I'm wondering what the preference is here and if there are other options I'm not considering:

1. Leave this as-is so we do not need to iterate over every data view every time the Step Define Rule component is rendered.
2. Push this down further into the Step Define Rule component (and even further still into the Data View Selector component) but need to generate the data view options array every time the Step Define Rule component (or if we decide to push it further down into the Data View Selector component) is rendered.

Open to suggestions here. My feeling was do it once when the parent component is rendered which leads me to believe maybe this should be in redux as well? 🤷‍♂️

Reply:
imo this should not be done in a component at all